multidownloader/evm_multidownloader.go

```diff
@@ -0,0 +1,607 @@
+// 	blockHeaders := []*aggkittypes.BlockHeader{aggkittypes.NewBlockHeaderFromEthHeader(blockHeader)}
+// 	err = dh.storage.SaveEthLogsWithHeaders(tx, blockHeaders, logs, false)
+// 	if err != nil {
+// 		return false, fmt.Errorf("Unsafe/Step: cannot save unsafe block: %w", err)
+// 	}
+// 	newSyncing := dh.pendingSync.UpdateSyncingAfterDoingQuery(logQueryData)
+// 	if err = dh.storage.UpdateSyncingStatus(tx, logQueryData); err != nil {
+// 		return false, fmt.Errorf("Unsafe/Step: cannot update syncing status: %w", err)
+// 	}
+// 	committed = true
+// 	if err := tx.Commit(); err != nil {
+// 		return false, fmt.Errorf("Unsafe/Step: cannot commit tx: %w", err)
+// 	}
+// 	dh.pendingSync = newSyncing
+// 	dh.log.Debugf("Unsafe/Step: finished block=%d syncing=%s",
+// 		blockHeader.Number.Uint64(),
+// 		dh.pendingSync.String())
+// 	err = dh.pendingSync.UpdateToBlock(ctx, dh.blockNotifierManager)
+// 	if err != nil {
+// 		return false, fmt.Errorf("Unsafe/Step: cannot update ToBlock in pendingSync: %w", err)
+// 	}
+// 	return dh.pendingSync.Finished(), nil
+// }
+
+func getBlockNumbers(logs []types.Log) []uint64 {
+	blockNumbers := make(map[uint64]struct{})
+	result := make([]uint64, 0)
+	for _, lg := range logs {
+		if _, exists := blockNumbers[lg.BlockNumber]; exists {
+			continue
+		}
+		blockNumbers[lg.BlockNumber] = struct{}{}
+		result = append(result, lg.BlockNumber)
+	}
+	return result
+}
+
+func (dh *EVMMultidownloader) IsAvailable(query mdrtypes.LogQuery) bool {
+	dh.mutex.Lock()
+	defer dh.mutex.Unlock()
+	return dh.syncedSegments.IsAvailable(query)
+}
+
+func mapBlockHeadersToList(blocks map[uint64]*aggkittypes.BlockHeader) []*aggkittypes.BlockHeader {
+	headers := make([]*aggkittypes.BlockHeader, 0, len(blocks))
+	for _, header := range blocks {
+		headers = append(headers, header)
+	}
+	return headers
+}
+
+// StepSafe performs a safe step syncing logs and block headers from historical data
+func (dh *EVMMultidownloader) StepSafe(ctx context.Context) (bool, error) {
+	logs, logQueryData, err := dh.requestLogs(ctx)
+	if err != nil {
+		if errors.Is(err, mdrtypes.ErrFinished) {
+			return true, nil
+		}
+		return false, fmt.Errorf("Safe/Step: cannot filter logs adapting block range: %w", err)
+	}
+	dh.log.Debugf("Safe/Step:: logs (%d) for blockRange=%s, addrs=%v", len(logs),
+		logQueryData.BlockRange.String(), logQueryData.Addrs)
+	blocks := getBlockNumbers(logs)
+	dh.log.Debugf("Safe/Step:: querying blockHeaders for %d blocks", len(blocks))
+	blockHeaders, err := etherman.RetrieveBlockHeaders(ctx, dh.log, dh.ethClient, dh.rpcClient,
+		blocks, dh.cfg.MaxParallelBlockHeaderRetrieval)
+	if err != nil {
+		return false, fmt.Errorf("Safe/Step: cannot retrieve block headers (%d): %w", len(blockHeaders), err)
+	}
+
+	// Calculate new state (not set in memory until commit is successful)
+	dh.mutex.Lock()
+	newSyncedSegments := dh.syncedSegments.Clone()
+	newPendingSegments := dh.pendingSync.Clone()
+	dh.mutex.Unlock()
```

Comment:
Hmmm, there is a lot of critical processing done after this lock, maybe it would be more safe to just hold the lock until we store the data in the `db`. This shouldn't take too much time to block any other routine. WDYT?